### ISSUE TITLE
MAINT: Move `sindex` to GeometryArray

### ISSUE THREAD

**Issue by adriangb** [geopandas/geopandas#1439]
Follow up to https://github.com/geopandas/geopandas/issues/1344#issuecomment-629784250. This would make sindex act like crs, in that there can be multiple geometry columns each with their own spatial index.

This would be done in 4 PRs:

1. Drop the `objects` parameter, make sindex operate only on `iloc/integer` indexes. Based on the fact that (1) this is not used internally, can easily be accomplished by using `iloc` and (3) feedback in https://github.com/shapely/shapely-rfc/pull/1#discussion_r428090169 seems to make sense.
2. Move `sindex` to `GeometryArray`.
3. Come up with an interface similar to `crs` to allow multiple geometry columns each with their own spatial index backend.
4. (Long term, after Shapely 2.0) allow users to specify their own spatial index implementation for each geometry column.

Aside from the `objects` parameter, no API changes would be needed: `GeoDataFrame.sindex` will point to `GeoDataFrame.geometry.sindex` with in turn will point to `GeoSeries.values.sindex` (which is `GeometryArray`).

This has some nice side effects, it gets rid of all of the `_invalidate_sindex` stuff in `GeoPandasBase`, and the input to the tree's `__init__` is more concise (i.e. always a numpy array of geometry objects).

**Comment by martinfleis** [comment]
> Drop the objects parameter, make sindex operate only on iloc/integer indexes. Based on the fact that (1) this is not used internally, (2) can easily be replaced by using iloc and (3) feedback in shapely/shapely-rfc#1 (comment) seems to make sense.

Can you explain again (or link the relevant comment) why are we removing it? The fact that it is not used internally is not relevant, it is a part of public API. 

Points 2 and 3 could be done together as it was done in case of CRS.

**Comment by adriangb** [comment]
The practical reason for removing it is that if we move `sindex` to `GeometryArray` we won't have the `loc/label` indexes to feed into the spatial index wrapper in the first place. Another practical reason is that it's implementation relies heavily on `rtree` and requires a hacky workaround in the `pygeos` spatial index wrapper. If we want to support more spatial index backends, we need a unified approach that minimally leverages features like this.

I would hardly say the `objects` parameter is part of the public API: 
1. `sindex.intersection` is not mentioned anywhere in the docs.
2. There is no signature for `intersection` in the old `sindex` (before we added pygeos) and if you called `help(GeoDataFrame.sindex.intersection)` you'd get back the rtree documentation.
3. If you do go and read the `rtree` documentation, all it says is that it allows storing arbitrary objects. You would then have to come back to geopandas, sift though the source code in `base.py` (where it lived until recently) and deduce that in geopandas, it is being used to store `loc/label` indexes. Or I guess you might be able to pick up on it from the tests. My point is that if someone was using that interface, they probable know what they are doing and won't be too phased if they have to manually re-index using `iloc`.

I think the advantages here outweigh the cons, but I'm open to discussion.

**Comment by martinfleis** [comment]
Okay, I am not against deprecating warning for `objects` in #1440. 

>>Points 2 and 3 could be done together as it was done in case of CRS.

>I'd love to do it together, I was just concerned it might be too big of a PR 😅

I know :). But I assume that it will not be a huge change, just pieces here and there. Once it will be stored on array level, then there's minimal work on 3. I even think that you have to do those two together, otherwise it will not work (e.g. switching active geometry).

**Comment by adriangb** [comment]
Cool! Let's get that deprecation warning into 0.8.0 and then we can work on the rest of it and see how it goes.

I tested moving it to the array level, it was actually _very_ straightforward. Mostly deleting `_invalidate_sindex()` calls from `base.py` and other places. I was a bit confused as to where/how to allow the users to specify their desired backend, but it's probably easier than it seems once I fully understand how the `crs` option works.

**Comment by adriangb** [comment]
Looking at this in a bit more depth, it looks to me like it will actually require a rather large refactor to implement the per-geometry column spatial index selection @martinfleis.

The main complication is the passing of the parameter between the different constructors and functions, ex for GeoDataFrame:
```python3
def set_geometry(self, col, drop=False, inplace=False, crs=None):
    # needs to be edited
def _ensure_geometry(data, crs=None):
    # needs to be edited
```
For GeoSeries, several methods including `_geoseries_constructor_with_fallback` need to have an `sindex_backend` parameter added.
Finally, within `GeometryArray` pretty much every every method needs to be edited:
```python3
def union(self, other):
        return GeometryArray(self._binary_method("union", self, other), crs=self.crs)
# needs to become something like
def union(self, other):
        return GeometryArray(self._binary_method("union", self, other), crs=self.crs, sindex_backend=self.sindex_backend)
```

I'm not sure what other `GeometryArray` properties might be added in the future, but maybe it would be easier to collect them under a set (like `GeometryArray.geometry_properties`) with the names of these parameters, similar to what scikit-learn does with [`get_params`](https://github.com/scikit-learn/scikit-learn/blob/fd237278e895b42abe8d8d09105cbb82dc2cbba7/sklearn/base.py#L189). That way methods like `set_geometry` could accept `**kwargs` and filter out the arguments that go to `GeometryArray` (`crs`, `sindex_backend`) from the rest (`inplace`, etc.).

**Comment by adriangb** [comment]
@martinfleis I updated the title to reflect the remaining work. My recollection of this situation is that we successfully moved sindex to GeometryArray, but we never implemented the ability to have per-column spatial index backends.

**Comment by martinfleis** [comment]
@adriangb isn't that a duplication of #1344 then? I'd be fine closing this as implemented in #1444 but if you think it is better to keep it as a separate issue, no problem with that either.
